### PR TITLE
ref: fix & simplify test callback

### LINKED ISSUE #4049: Better/correct test for callback
## 🚀 Feature

Reopen #4009 and let it merge...

### Motivation

The actual test with bools is weak, do not check call order

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

simplify the test for a general callback so now we can also see the order

Resolves #4049

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
I would prefer to do this with unittest.Mock, and then for example:
```python
callback.on_batch_start.assert_called_with(...) 
or 
callback.on_batch_start.assert_called_once()
or 
callback.on_batch_start.call_count == ...
```
it is quite elegant

**Comment by Borda** [comment]
> it is quite elegant and opens a lot more options for rigorous testing. we can even test for the order in which these are called, or with which arguments etc.

can we also see in which order the methods were called? I know that we can be easily counted, but each mock is independent, right?

**Comment by awaelchli** [comment]
Yes, absolutely!
See my example here: https://github.com/PyTorchLightning/pytorch-lightning/issues/4049#issuecomment-706598291

File: tests/callbacks/test_callbacks.py [hunk 1]
```diff
@@ -16,251 +18,193 @@ def _check_args(trainer, pl_module):
...
-            self.on_test_start_called = False
-            self.on_test_end_called = False
+            self.called = []
 
         def setup(self, trainer, pl_module, stage: str):
+            self.called.append(inspect.currentframe().f_code.co_name)
             assert isinstance(trainer, Trainer)
-            self.setup_called = True
 
         def teardown(self, trainer, pl_module, step: str):
```
Comment (HanSilver01) [inline tests/callbacks/test_callbacks.py:50]: [[]()]()

Comment (HanSilver01) [reply tests/callbacks/test_callbacks.py:50]: Rico de verte

File: tests/callbacks/test_callbacks.py [hunk 2]
```diff
@@ -16,251 +18,193 @@ def _check_args(trainer, pl_module):
...
+        'on_pretrain_routine_start',
+        'on_pretrain_routine_end',
+        'on_test_start',
+        'on_test_batch_start',
+        'on_test_batch_end',
+        'on_test_batch_start',
+        'on_test_batch_end',
+        'on_test_end',
+        'on_fit_end',
+        'teardown',
```
Comment (HanSilver01) [inline tests/callbacks/test_callbacks.py:208]: ![Cargando 99869690...]

**Comment by Borda** [comment]
> it is quite elegant and opens a lot more options for rigorous testing. we can even test for the order in which these are called, or with which arguments etc.

mind share example how would you test the order of calling?  @awaelchli :]

**Comment by awaelchli** [comment]
with `assert_has_calls`, see example here: https://github.com/PyTorchLightning/pytorch-lightning/issues/4049#issuecomment-706598291

**Comment by Borda** [comment]
> with `assert_has_calls`, see example here: [#4049 (comment)](https://github.com/PyTorchLightning/pytorch-lightning/issues/4049#issuecomment-706598291)

ok, I needed to find a way around their native asserts but shall be fine now :] @awaelchli

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by SeanNaren — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
very nice! Thanks for taking my advice @Borda

**Pull request merged**